### PR TITLE
patch calico ip detection method before applying calico manifest

### PR DESCRIPTION
**Pull request by devalpatel67**

### CODE REVIEW

File: k3sAndProviderServices.sh [hunk 1]
```diff
@@ -153,6 +153,18 @@ update_coredns_config() {
...
+    # Ensure jq is installed for JSON processing
+    if ! command -v jq &> /dev/null; then
+        echo "jq is not installed. Installing jq..."
+        apt-get update && apt-get install -y jq
+    fi
+
+    # Ensure yq is installed for YAML processing
+    if ! command -v yq &> /dev/null; then
+        echo "yq is not installed. Installing yq..."
+        sudo curl -L https://github.com/mikefarah/yq/releases/latest/download/yq_linux_amd64 -o /usr/bin/yq && chmod +x /usr/bin/yq
```
Comment (ghost) [inline k3sAndProviderServices.sh]: Please remove "sudo".
The script is supposed to be run as root anyway.

Comment (devalpatel67) [reply k3sAndProviderServices.sh]: Done. Thanks for pointing it out.

File: k3sAndProviderServices.sh [hunk 2]
```diff
@@ -153,6 +153,18 @@ update_coredns_config() {
...
+    # Ensure jq is installed for JSON processing
+    if ! command -v jq &> /dev/null; then
+        echo "jq is not installed. Installing jq..."
+        apt-get update && apt-get install -y jq
+    fi
+
+    # Ensure yq is installed for YAML processing
+    if ! command -v yq &> /dev/null; then
+        echo "yq is not installed. Installing yq..."
+        curl -L https://github.com/mikefarah/yq/releases/latest/download/yq_linux_amd64 -o /usr/bin/yq && chmod +x /usr/bin/yq
```
Comment (ghost) [inline k3sAndProviderServices.sh]: let's install external components under `/usr/local/bin/` instead

**Comment by ghost** [comment]
It looks good to me.
I had to normalize the yaml before diff'ing it, e.g.:

```
curl -O https://raw.githubusercontent.com/projectcalico/calico/refs/tags/v3.28.2/manifests/calico.yaml
cat calico.yaml | yq . > calico.yaml.bef.normalized
```

File: k3sAndProviderServices.sh (hunk 2 above)
Comment (devalpatel67) [reply k3sAndProviderServices.sh]: Done.

**Comment by devalpatel67** [comment]
> It looks good to me. I had to normalize the yaml before diff'ing it, e.g.:
> 
> ```
> curl -O https://raw.githubusercontent.com/projectcalico/calico/refs/tags/v3.28.2/manifests/calico.yaml
> cat calico.yaml | yq . > calico.yaml.bef.normalized
> ```

Normalization mainly affects the `description` attribute. Do we need to normalize it after downloading it every time, or is it just for your different purposes?

**Comment by ghost** [comment]
> > It looks good to me. I had to normalize the yaml before diff'ing it, e.g.:
> > ```
> > curl -O https://raw.githubusercontent.com/projectcalico/calico/refs/tags/v3.28.2/manifests/calico.yaml
> > cat calico.yaml | yq . > calico.yaml.bef.normalized
> > ```
> 
> Normalization mainly affects the `description` attribute. Do we need to normalize it after downloading it every time, or is it just for your different purposes?

Just for me ;) wanted to make sure yq didn't break anything.

**Review by ghost — APPROVED** [review]

**Comment by ghost** [comment]
@chainzero it is ready to merge.

**Pull request merged**